Mathlib/Data/Fin/Basic.lean

```diff
@@ -104,231 +212,2385 @@ lemma Fin.ofNat'_succ : {n : Nat} → [Nonempty (Fin n)] →
+-/
+@[elab_as_elim]
+def reverseInduction {C : Fin (n + 1) → Sort _} (hlast : C (Fin.last n))
+    (hs : ∀ i : Fin n, C i.succ → C (castSucc i)) (i : Fin (n + 1)) : C i :=
+  if hi : i = Fin.last n then _root_.cast (congr_arg C hi.symm) hlast
+  else
+    let j : Fin n := ⟨i, lt_of_le_of_ne (Nat.le_of_lt_succ i.2) fun h => hi (Fin.ext h)⟩
+    have : n - i < n + 1 - i :=
+      lt_of_eq_of_lt (Nat.add_sub_add_right ..).symm
+        (Nat.sub_lt_sub_left i.2 (Nat.lt_succ_self i))
+    have hi : i = Fin.castSucc j := Fin.ext rfl
+    _root_.cast (congr_arg C hi.symm) (hs _ (reverseInduction hlast hs j.succ))
+termination_by _ => n + 1 - i
+#align fin.reverse_induction Fin.reverseInduction
+
+@[simp]
+theorem reverse_induction_last {n : ℕ} {C : Fin (n + 1) → Sort _} (h0 : C (Fin.last n))
+    (hs : ∀ i : Fin n, C i.succ → C (castSucc i)) :
+    (reverseInduction h0 hs (Fin.last n) : C (Fin.last n)) = h0 := by
+  rw [reverseInduction] ; simp
+#align fin.reverse_induction_last Fin.reverse_induction_last
+
+@[simp]
+theorem reverse_induction_cast_succ {n : ℕ} {C : Fin (n + 1) → Sort _} (h0 : C (Fin.last n))
+    (hs : ∀ i : Fin n, C i.succ → C (castSucc i)) (i : Fin n) :
+    (reverseInduction h0 hs (castSucc i) : C (castSucc i)) = hs i (reverseInduction h0 hs i.succ) :=
+  by
+  rw [reverseInduction, dif_neg (_root_.ne_of_lt (Fin.castSucc_lt_last i))]
+  cases i
+  rfl
+#align fin.reverse_induction_cast_succ Fin.reverse_induction_cast_succ
+
+/-- Define `f : Π i : fin n.succ, C i` by separately handling the cases `i = fin.last n` and
+`i = j.cast_succ`, `j : fin n`. -/
+@[elab_as_elim]
+def lastCases {n : ℕ} {C : Fin (n + 1) → Sort _} (hlast : C (Fin.last n))
+    (hcast : ∀ i : Fin n, C (castSucc i)) (i : Fin (n + 1)) : C i :=
+  reverseInduction hlast (fun i _ => hcast i) i
+#align fin.last_cases Fin.lastCases
+
+@[simp]
+theorem last_cases_last {n : ℕ} {C : Fin (n + 1) → Sort _} (hlast : C (Fin.last n))
+    (hcast : ∀ i : Fin n, C (castSucc i)) :
+    (Fin.lastCases hlast hcast (Fin.last n) : C (Fin.last n)) = hlast :=
+  reverse_induction_last _ _
+#align fin.last_cases_last Fin.last_cases_last
+
+@[simp]
+theorem last_cases_cast_succ {n : ℕ} {C : Fin (n + 1) → Sort _} (hlast : C (Fin.last n))
+    (hcast : ∀ i : Fin n, C (castSucc i)) (i : Fin n) :
+    (Fin.lastCases hlast hcast (Fin.castSucc i) : C (Fin.castSucc i)) = hcast i :=
+  reverse_induction_cast_succ _ _ _
+#align fin.last_cases_cast_succ Fin.last_cases_cast_succ
+
+/-- Define `f : Π i : fin (m + n), C i` by separately handling the cases `i = cast_add n i`,
+`j : fin m` and `i = nat_add m j`, `j : fin n`. -/
+@[elab_as_elim]
+def addCases {m n : ℕ} {C : Fin (m + n) → Sort u} (hleft : ∀ i, C (castAdd n i))
+    (hright : ∀ i, C (natAdd m i)) (i : Fin (m + n)) : C i :=
+  if hi : (i : ℕ) < m then Eq.recOn (cast_add_cast_lt n i hi) (hleft (castLt i hi))
+  else Eq.recOn (natAdd_subNat_cast (le_of_not_lt hi)) (hright _)
```

Comment:
I am a bit confused by the state of this file. On my local machine if I pull and compile, then I get an error in this declaration, with Eq.recOn twice complaining about the motive. It's possible to supply the motives explicitly. Here's a fix where I supply the correct motive once, and use the non-stupid triangle (it's got much cleverer in Lean 4) for the other error:

```lean
@[elab_as_elim]
def addCases {m n : ℕ} {C : Fin (m + n) → Sort u} (hleft : ∀ i, C (castAdd n i))
    (hright : ∀ i, C (natAdd m i)) (i : Fin (m + n)) : C i :=
  if hi : (i : ℕ) < m then Eq.recOn 
    (motive := λ a (_ : ((castAdd n).toEmbedding (castLt i hi)) = a) => C a)
    (cast_add_cast_lt n i hi) (hleft (castLt i hi))
  else (natAdd_subNat_cast (le_of_not_lt hi)) ▸ (hright _)
```

But in fact you can check that this is defeq to
```lean
@[elab_as_elim]
def addCases' {m n : ℕ} {C : Fin (m + n) → Sort u} (hleft : ∀ i, C (castAdd n i))
    (hright : ∀ i, C (natAdd m i)) (i : Fin (m + n)) : C i :=
  if hi : (i : ℕ) < m then cast_add_cast_lt n i hi ▸ hleft (castLt i hi)
  else (natAdd_subNat_cast (le_of_not_lt hi)) ▸ (hright _)
```
so if there is actually an error here then this is my proposed fix, and if there isn't an error here then please feel free to ignore me.